qutip/tests/test_states.py

```diff
@@ -66,63 +69,124 @@ def test_implicit_tensor_projection(size, n, m):
-            assert_(rho.data[i, i] == 1.0)
-
-    def testTripletStateNorm(self):
-        """
-        Test the states returned by function triplet_states are normalized.
-        """
-        for triplet in triplet_states():
-            assert_(abs(triplet.norm() - 1.) < 1e-12)
-
-
-if __name__ == "__main__":
-    run_module_suite()
+@pytest.mark.parametrize("base, operator, args, opargs, eigenval", [
+    pytest.param(qutip.basis, qutip.num, (10, 3), (10,), 3,
+                 id="basis"),
+    pytest.param(qutip.basis, qutip.num, (10, 3, 1), (10, 1), 3,
+                 id="basis_offset"),
+    pytest.param(qutip.fock, qutip.num, (10, 3), (10,), 3,
+                 id="fock"),
+    pytest.param(qutip.fock_dm, qutip.num, (10, 3), (10,), 3,
+                 id="fock_dm"),
+    pytest.param(qutip.fock_dm, qutip.num, (10, 3, 1), (10, 1), 3,
+                 id="fock_dm_offset"),
+    pytest.param(qutip.coherent, qutip.destroy, (20, 0.75), (20,), 0.75,
+                 id="coherent"),
+    pytest.param(qutip.coherent, qutip.destroy, (50, 4.25, 1), (50, 1), 4.25,
+                 id="coherent_offset"),
+    pytest.param(qutip.coherent_dm, qutip.destroy, (25, 1.25), (25,), 1.25,
+                 id="coherent_dm"),
+    pytest.param(qutip.phase_basis, qutip.phase, (10, 3), (10,), 3 * _2PI / 10,
+                 id="phase_basis"),
+    pytest.param(qutip.phase_basis, qutip.phase,
+                 (10, 3, 1), (10, 1), 3 * _2PI / 10 + 1,
+                 id="phase_basis_phi0"),
+    pytest.param(qutip.spin_state, qutip.spin_Jz, (3, 2), (3,), 2,
+                 id="spin_state"),
+    pytest.param(qutip.zero_ket, qutip.qeye, (10,), (10,), 0,
+                 id="zero_ket"),
+])
+def test_diverse_basis(base, operator, args, opargs, eigenval):
+    # For state which are supposed to eigenvector of an operator
+    # Verify that correspondance
+    state = base(*args)
+    oper = operator(*opargs)
+    assert qutip.expect(oper, state) == pytest.approx(eigenval)
+
+
+@pytest.mark.parametrize('dm', [
+    partial(qutip.thermal_dm, n=1.),
+    qutip.maximally_mixed_dm,
+    partial(qutip.coherent_dm, alpha=0.5),
+    partial(qutip.fock_dm, n=1),
+    partial(qutip.spin_state, m=2, type='dm'),
+    partial(qutip.spin_coherent, theta=1, phi=2, type='dm'),
+], ids=[
+    'thermal_dm', 'maximally_mixed_dm', 'coherent_dm',
+    'fock_dm', 'spin_state', 'spin_coherent'
+])
+def test_dm(dm):
+    N = 5
+    rho = dm(N)
+    # make sure rho has trace close to 1.0
+    assert abs(rho.tr() - 1.0) < 1e-12
+
+
+def test_CoherentState():
+    N = 10
+    alpha = 0.5
+    c1 = qutip.coherent(N, alpha)  # displacement method
+    c2 = qutip.coherent(7, alpha, offset=3)  # analytic method
+    assert abs(qutip.expect(qutip.destroy(N), c1) - alpha) < 1e-10
+    assert (qutip.Qobj(c1[3:]) - c2).norm() < 1e-7
+
+
+def test_TripletStateNorm():
+    for triplet in qutip.triplet_states():
+        assert abs(triplet.norm() - 1.) < 1e-12
```

Comment:
There's presumably more we'd like to test about these states, right?  Should we test they're orthonormal, and check their eigenvalues?  Also, the documentation for the function seems to be wrong (although the code isn't) - one of the documentated states is the singlet, not a triplet!!